README
== Description
  A simple tar & compress library that nicely wraps external system calls.

== Installation
  gem install archive-tar-external
   
== Synopsis
  require 'archive/tar/external'
  include Archive
   
  # The long way
  t = Tar::External.new('test.tar')
  t.create_archive('*.rb')
  t.compress_archive('gzip')
   
  # The short way
  t = Tar::External.new('test.tar', '*.rb', 'gzip')

== Prerequisites
  The 'tar' command line program.
  At least one compression program, e.g. gzip, bzip2, zip, etc.
   
== Known Issues
  The tar program that comes with Solaris will not raise an error if you
  try to expand a file from an archive that does not contain that file.

  If you come across any other issues, please report them on the project
  page at https://github.com/djberg96/archive-tar-external.

== History
  This project was originally named "archive-tarsimple", but was renamed
  on April 7, 2006. Versions of this gem prior to that date are no longer
  available.

== License
  Apache-2.0

== Warranty
  This package is provided "as is" and without any express or
  implied warranties, including, without limitation, the implied
  warranties of merchantability and fitness for a particular purpose.

== Copyright
  (C) 2003 - 2019 Daniel J. Berger
  All Rights Reserved

== Author
  Daniel J. Berger


CHANGES
== 1.4.0 - 23-Jan-2019
* Changed license to Apache-2.0.
* The VERSION constant is now frozen.
* Added some metadata to the gemspec.
* Cert updated, should be good for about 10 years.

== 1.3.3 - 16-Jan-2016
* This gem is now signed.
* Fixed some deprecation warnings in the test suite.
* Added an archive-tar-external.rb file for convenience.
* Updates to the Rakefile.

== 1.3.2 - 14-Mar-2014
* The create_archive method now accepts an optional second argument that
  specifies options to the tar command. The default is 'cf' (create file).
* The gem:create task was updated for Rubygems 2.x.
* Added rake as a development dependency.
* Ruby 1.8.x support has been dropped. This only matters for MS Windows.

== 1.3.1 - 20-Sep-2011
* Refactored the Rakefile a bit.

== 1.3.0 - 16-Jan-2010
* Altered the source code layout, which requires a change in your require
  line. It's now require 'archive/tar/external'.
* Moved some content from the .txt file to the README.
* Updated one test to use a skip as needed.
* Moved source code to github.

== 1.2.3 - 25-Sep-2009
* Fixed a packaging bug.
* Added cygwin and mingw to Windows checks.
* Now requires open3 properly for Ruby 1.9 on Windows.
* Minor modifications to the gemspec.
* Added the 'gem' rake task.

== 1.2.2 - 28-Jul-2009
* Compatibility fixes for Ruby 1.9.x and JRuby.
* Singleton methods that were previously synonyms are now true aliases.
* License changed to Artistic 2.0.
* Several gemspec updates, including an updated description, the addition
  of a license and development dependencies and refactored platform handling.
* Added test-unit and ptools as development dependencies for the test suite.
* Some refactoring of the test suite to take advantage of some of the features
  in test-unit 2.x.
* Renamed the test file to test_archive_tar_external.rb.

== 1.2.1 - 30-Jul-2007
* The TarError class is now Tar::Error.
* Added a Rakefile with tasks for installation and testing.
* Removed the install.rb file. Installation is now handled by the Rakefile.
* Documentation updates.

== 1.2.0 - 7-Apr-2006
* Project renamed 'archive-tar-external'.
* The 'Tar' class is now Tar::External.
* The RAA project name is now tar-external.
* Added the Tar::External.uncompress_archive class method for archives
  that already exist.
* Added the Tar::External.extract_archive class method for tarballs
  that already exist.
* Added the Tar::External#compressed_archive_name= method.

== 1.1.1 - 1-Mar-2006
* Replaced PLATFORM with RUBY_PLATFORM since the former is deprecated
  for Ruby 1.9/2.0.

== 1.1.0 - 10-Feb-2006
* Internal refactoring changes - now uses block form of Open3.popen3.
* Most methods now return self instead of true (better for chaining).
* Corresponding test suite changes.

== 1.0.0 - 1-Sep-2005
* Moved project to RubyForge.
* Modified the constructor to accept a file pattern.  If present, becomes
  a shortcut for Tar.new + Tar#create_archive.
* Constructor merely calls .to_s on the archive name, rather than validating
  that it's a String argument.
* You can now specify the tar program you wish to use, e.g. "tar", "gtar", etc.
* Fixed a bug in Tar#uncompress_archive where it wasn't finding the appropriate
  compressed archive name.
* Added aliases for most methods whereby you can omit the word "archive" in
  the method name now, e.g. Tar#compress is an alias for Tar#compress_archive.
* Now more polite about closing open handles from Open3.
* The (rdoc) documentation is now inlined.
* Removed the tarsimple.rd and tarsimple.html files.
* Made most documentation rdoc friendly.
* Updated the README and MANIFEST files.
* Removed the INSTALL file.  The installation instructions are now included
  in the README file.
* Test suite updates.
* Added a gemspec.

== 0.3.0 - 10-Aug-2004
* Added the update_archive() method.
* The expand_archive() method has been renamed to extract_archive().
  For backward's compatability, an alias remains so you may use
  either method name.
* The extract_archive() method now accepts a list of file names as
  an argument.  If present, then only those file names are extracted.
  If not, then the entire archive is extracted.
* Added corresponding tests and documentation updates.

== 0.2.0 - 9-Aug-2004
* Added the add_to_archive() method.
* Removed the VERSION() class method. Use the constant instead.
* Changed "TarException" and "CompressionException" to "TarError" and
  "CompressError", respectively.
* Moved rd documentation into its own file under the 'doc' directory.
* Added unit tests and updated the docs.
* Removed the html file.  You can generate that on your own if you
  wish using the rd2 tool.
* Added warranty information.

== 0.1.0 - 24-Jan-2003
* Initial Release


test/test_archive_tar_external.rb
###############################################################################
# test_archive_tar_external.rb
#
# Test suite for the archive-tar-external library. This test case should be
# run via the 'rake test' Rake task.
###############################################################################
require 'archive/tar/external'
require 'test-unit'
require 'ptools'
include Archive

class TC_ArchiveTarExternal < Test::Unit::TestCase
  def self.startup
    Dir.chdir(File.dirname(File.expand_path(__FILE__)))

    @@tmp_file1 = 'temp1.txt'
    @@tmp_file2 = 'temp2.txt'
    @@tmp_file3 = 'temp3.txt'

    @@gtar_found  = File.which('gtar')
    @@tar_found   = File.which('tar')
    @@gzip_found  = File.which('gzip')
    @@bzip2_found = File.which('bzip2')

    File.open(@@tmp_file1, 'w'){ |f| f.puts 'This is a temporary text file' }
    File.open(@@tmp_file2, 'w'){ |f| f.puts 'This is a temporary text file' }
    File.open(@@tmp_file3, 'w'){ |f| f.puts 'This is a temporary text file' }
  end

  def setup
    @tar      = Tar::External.new('test.tar')
    @tar_name = 'test.tar'
    @pattern  = '*.txt'
    @archive  = 'temp.tar.gz'
  end

  def test_version
    assert_equal('1.4.0', Tar::External::VERSION)
    assert_true(Tar::External::VERSION.frozen?)
  end

  def test_constructor
    assert_nothing_raised{ Tar::External.new(@tar_name) }
  end

  def test_constructor_with_extension
    assert_nothing_raised{ Tar::External.new(@tar_name, '*.txt') }
  end

  def test_constructor_with_program
    omit_unless(@@gzip_found){ 'gzip program not found - skipping' }
    assert_nothing_raised{ Tar::External.new(@tar_name, '*.txt', 'gzip') }
  end

  def test_constructor_expected_errors
    assert_raise(ArgumentError){ Tar::External.new }
  end

  def test_tar_program
    assert_respond_to(@tar, :tar_program)
    assert_equal('tar', @tar.tar_program)
  end

  def test_archive_name
    assert_respond_to(@tar, :archive_name)
    assert_respond_to(@tar, :archive_name=)

    assert_equal('test.tar', @tar.archive_name)
    assert_nothing_raised{ @tar.archive_name }
    assert_nothing_raised{ @tar.archive_name = 'foo' }
  end

  def test_compressed_archive_name_get
    assert_respond_to(@tar, :compressed_archive_name)
    assert_nil(@tar.compressed_archive_name)
  end

  def test_compressed_archive_name_set
    assert_respond_to(@tar, :compressed_archive_name=)
    assert_nothing_raised{ @tar.compressed_archive_name = 'test.tar.gz' }
    assert_equal('test.tar.gz', @tar.compressed_archive_name)
    assert_equal('test.tar', @tar.archive_name)

    assert_nothing_raised{ @tar.compressed_archive_name = 'test.tgz' }
    assert_equal('test.tgz', @tar.compressed_archive_name)
    assert_equal('test.tar', @tar.archive_name)
  end

  test "create_archive basic functionality" do
    assert_respond_to(@tar, :create_archive)
    assert_nothing_raised{ @tar.create_archive(@pattern) }
    assert_true(File.exist?(@tar_name))
  end

  test "create_archive requires at least on argument" do
    assert_raises(ArgumentError){ @tar.create_archive }
  end

  test "create_archive raises an error if no files match the pattern" do
    assert_raises(Tar::Error){ @tar.create_archive('*.blah') }
  end

  test "create_archive accepts optional parameters" do
    assert_nothing_raised{ @tar.create_archive(@pattern, 'cfj') }
  end

  def test_create_alias
    assert_respond_to(@tar, :create)
    assert_true(Tar::External.instance_method(:create) == Tar::External.instance_method(:create_archive))
  end

  def test_compress_archive_basic
    assert_respond_to(@tar, :compress_archive)
  end

  def test_compress_alias
    assert_respond_to(@tar, :compress)
    assert_true(Tar::External.instance_method(:compress) == Tar::External.instance_method(:compress_archive))
  end

  def test_compress_archive_gzip
    assert_nothing_raised{ @tar.create_archive('*.txt') }
    assert_nothing_raised{ @tar.compress_archive }

    assert_equal('test.tar.gz', @tar.compressed_archive_name)
    assert_true(File.exist?('test.tar.gz'))
  end

  def test_compress_archive_bzip2
    assert_nothing_raised{ @tar.create_archive('*.txt') }
    assert_nothing_raised{ @tar.compress_archive('bzip2') }
    assert_true(File.exist?('test.tar.bz2'))
  end

  def test_uncompress_archive
    assert_respond_to(@tar, :uncompress_archive)
    assert_nothing_raised{ @tar.create_archive('*.txt') }
    assert_nothing_raised{ @tar.compress_archive }
    assert_nothing_raised{ @tar.uncompress_archive }
    assert_false(File.exist?('test.tar.gz'))
  end

  def test_uncompress_archive_class_method
    assert_respond_to(Tar::External, :uncompress_archive)
  end

  def test_uncompress_alias
    assert_respond_to(Tar::External, :uncompress)
    assert_true(Tar::External.method(:uncompress) == Tar::External.method(:uncompress_archive))
  end

  def test_archive_info
    assert_respond_to(@tar, :archive_info)
    assert_nothing_raised{ @tar.create_archive('*.txt') }
    assert_equal(['temp1.txt','temp2.txt','temp3.txt'], @tar.archive_info)
  end

  def test_add_to_archive
    assert_respond_to(@tar,:add_to_archive)
    assert_nothing_raised{ @tar.create_archive('temp1.txt') }
    assert_nothing_raised{ @tar.add_to_archive('temp2.txt') }
    assert_nothing_raised{ @tar.add_to_archive('temp2.txt','temp3.txt') }
  end

  def test_update_archive
    assert_respond_to(@tar, :update_archive)
    assert_nothing_raised{ @tar.create_archive('*.txt') }
    assert_nothing_raised{ @tar.update_archive('temp2.txt') }
  end

  def test_extract_archive_basic
    assert_respond_to(@tar, :extract_archive)
  end

  def test_extract_archive_aliases
    assert_true(Tar::External.instance_method(:extract_archive) == Tar::External.instance_method(:expand_archive))
    assert_true(Tar::External.instance_method(:extract) == Tar::External.instance_method(:expand_archive))
    assert_true(Tar::External.instance_method(:expand) == Tar::External.instance_method(:expand_archive))
  end

  def test_extract_archive_advanced
    omit_unless(RbConfig::CONFIG['host_os'] =~ /sunos|solaris/){
      assert_nothing_raised{ @tar.tar_program = @@gtar }
    }
    assert_nothing_raised{ @tar.create('*.txt') }
    assert_raises(Tar::Error){ @tar.expand('blah.txt') }

    assert_nothing_raised{ @tar.extract_archive }
    assert_nothing_raised{ @tar.extract_archive('temp2.txt') }
  end

  def teardown
    @tar = nil
    File.delete('test.tar') if File.exist?('test.tar')
    File.delete('test.tar.gz') if File.exist?('test.tar.gz')
    File.delete('test.tar.bz2') if File.exist?('test.tar.bz2')
    File.delete('test.tar.zip') if File.exist?('test.tar.zip')
  end

  def self.shutdown
    @@tar_foudn   = nil
    @@gzip_found  = nil
    @@bzip2_found = nil

    File.delete(@@tmp_file1) if File.exist?(@@tmp_file1)
    File.delete(@@tmp_file2) if File.exist?(@@tmp_file2)
    File.delete(@@tmp_file3) if File.exist?(@@tmp_file3)
  end
end


lib/archive/tar/external.rb
require 'open3'

# The Archive module serves as a namespace only.
module Archive

  # The Tar class serves as a toplevel class namespace only.
  class Tar

    # Raised if something goes wrong during the execution of any methods
    # which use the tar command internally.
    class Error < StandardError; end

    # Raised if something goes wrong during the Tar#compress_archive or
    # Tar#uncompress_archive methods.
    class CompressError < StandardError; end

    # This class encapsulates tar & zip operations.
    class Tar::External
      # The version of the archive-tar-external library.
      VERSION = '1.4.0'.freeze

      # The name of the archive file to be used, e.g. "test.tar"
      attr_accessor :archive_name

      # The name of the tar program you wish to use.  The default is "tar".
      attr_accessor :tar_program

      # The name of the archive file after compression, e.g. "test.tar.gz"
      attr_reader :compressed_archive_name

      # Returns an Archive::Tar::External object.  The +archive_name+ is the
      # name of the tarball.  While a .tar extension is recommended based on
      # years of convention, it is not enforced.
      #
      # Note that this does not actually create the archive unless you
      # pass a value to +file_pattern+.  This then becomes a shortcut for
      # Archive::Tar::External.new + Archive::Tar::External#create_archive.
      #
      # If +program+ is provided, then it compresses the archive as well by
      # calling Archive::Tar::External#compress_archive internally.
      #
      def initialize(archive_name, file_pattern=nil, program=nil)
        @archive_name            = archive_name.to_s
        @compressed_archive_name = nil
        @tar_program             = 'tar'

        if file_pattern
          create_archive(file_pattern)
        end

        if program
          compress_archive(program)
        end
      end

      # Assign a compressed archive name.  This autogenerates the archive_name
      # based on the extension of the name provided, unless you provide the
      # extension yourself.  If the extension is '.tgz', then the base of the
      # name + '.tar' will be the new archive name.
      #
      # This should only be used if you have a pre-existing, compressed archive
      # that you want to uncompress, and want to have a Tar::External object
      # around.  Otherwise, use the class method Tar::External.uncompress.
      #
      def compressed_archive_name=(name, ext=File.extname(name))
        if ext.downcase == '.tgz'
          @archive_name = File.basename(name, ext.downcase) + '.tar'
        else
          @archive_name = File.basename(name, ext)
        end
        @compressed_archive_name = name
      end

      # Creates the archive using +file_pattern+ using +options+ or 'cf'
      # (create file) by default.
      #
      # Raises an Archive::Tar::Error if a failure occurs.
      #
      def create_archive(file_pattern, options = 'cf')
        cmd = "#{@tar_program} #{options} #{@archive_name} #{file_pattern}"

        Open3.popen3(cmd){ |tar_in, tar_out, tar_err|
          err = tar_err.gets
          raise Error, err.chomp if err
        }

        self
      end

      alias :create :create_archive

      # Compresses the archive with +program+, or gzip if no program is
      # provided.  If you want to pass arguments to +program+, merely include
      # them as part of the program name, e.g. "gzip -f".
      #
      # Any errors that occur here will raise a Tar::CompressError.
      #
      def compress_archive(program='gzip')
        cmd = "#{program} #{@archive_name}"

        Open3.popen3(cmd){ |prog_in, prog_out, prog_err|
          err = prog_err.gets
          raise CompressError, err.chomp if err

          # Find the new file name with the extension.  There's probably a more
          # reliable way to do this, but this should work 99% of the time.
          name = Dir["#{@archive_name}.{gz,bz2,cpio,zip}"].first
          @compressed_archive_name = name
        }

        self
      end

      alias :compress :compress_archive

      # Uncompresses the tarball using the program you pass to this method.  The
      # default is "gunzip".  Just as for +compress_archive+, you can pass
      # arguments along as part of the argument.
      #
      # Note that this is only for use with archives that have been zipped up
      # with gunzip, or whatever.  If you want to *extract* the files from the
      # tarball, use Tar::External#extract instead.
      #
      # Any errors that occur here will raise a Tar::CompressError.
      #
      def uncompress_archive(program="gunzip")
        unless @compressed_archive_name
          raise CompressError, "no compressed file found"
        end

        cmd = "#{program} #{@compressed_archive_name}"

        Open3.popen3(cmd){ |prog_in, prog_out, prog_err|
          err = prog_err.gets
          raise CompressError, err.chomp if err
          @compressed_archive_name = nil
        }
        self
      end

      alias :uncompress :uncompress_archive

      # Uncompress an existing archive, using +program+ to uncompress it.
      # The default decompression program is gunzip.
      #
      def self.uncompress_archive(archive, program='gunzip')
        cmd = "#{program} #{archive}"

        Open3.popen3(cmd){ |prog_in, prog_out, prog_err|
          err = prog_err.gets
          raise CompressError, err.chomp if err
        }
      end

      class << self
        alias uncompress uncompress_archive
      end

      # Returns an array of file names that are included within the tarball.
      # This method does not extract the archive.
      #
      def archive_info
        result = []
        cmd = "#{@tar_program} tf #{@archive_name}"
        Open3.popen3(cmd){ |ain, aout, aerr|
          err = aerr.gets
          raise Error, err.chomp if err

          while output = aout.gets
            result << output.chomp
          end
        }
        result
      end

      alias :info :archive_info

      # Adds +files+ to an already existing archive.
      #
      def add_to_archive(*files)
        if files.empty?
          raise Error, "there must be at least one file specified"
        end

        cmd = "#{@tar_program} rf #{@archive_name} #{files.join(" ")}"

        Open3.popen3(cmd){ |ain, aout, aerr|
          err = aerr.gets
          raise Error, err.chomp if err
        }
        self
      end

      alias :add :add_to_archive

      # Updates the given +files+ in the archive, i.e. they are added if they
      # are not already in the archive or have been modified.
      #
      def update_archive(*files)
        if files.empty?
          raise Error, "there must be at least one file specified"
        end

        cmd = "#{@tar_program} uf #{@archive_name} #{files.join(" ")}"

        Open3.popen3(cmd){ |ain, aout, aerr|
          err = aerr.gets
          raise Error, err.chomp if err
        }

        self
      end

      alias :update :update_archive

      # Expands the contents of the tarball.  It does NOT delete the tarball.
      # If +files+ are provided, then only those files are extracted.
      # Otherwise, all files are extracted.
      #
      # Note that some tar programs, notably the tar program shipped by Sun,
      # does not issue any sort of warning or error if you try to extract a
      # file that does not exist in the archive.
      #
      def extract_archive(*files)
        cmd = "#{@tar_program} xf #{@archive_name}"

        unless files.empty?
          cmd << " " << files.join(" ")
        end

        Open3.popen3(cmd){ |ain, aout, aerr|
          err = aerr.gets
          raise Error, err.chomp if err
        }

        self
      end

      alias :expand_archive :extract_archive
      alias :extract :extract_archive
      alias :expand :extract_archive

      # A class method that behaves identically to the equivalent instance
      # method, except that you must specifiy that tarball as the first
      # argument. Also, the tar program is hard coded to 'tar xf'.
      #
      def self.extract_archive(archive, *files)
        cmd = "tar xf #{archive}"

        unless files.empty?
          cmd << " " << files.join(" ")
        end

        Open3.popen3(cmd){ |ain, aout, aerr|
          err = aerr.gets
          raise Error, err.chomp if err
        }

        self
      end

      class << self
        alias expand_archive extract_archive
        alias extract extract_archive
        alias expand extract_archive
      end
    end
  end
end


lib/archive/tar_external.rb
require 'rbconfig'

if Config::CONFIG['host_os'] =~ /mswin|dos|win32|cygwin|mingw/i
   if RUBY_VERSION.to_f < 1.9 && RUBY_PLATFORM !~ /java/i
      require 'win32/open3'
   else
      require 'open3'
   end
else
   require 'open3'
end

# The Archive module serves as a namespace only.
module Archive

   # The Tar class serves as a toplevel class namespace only.
   class Tar

      # Raised if something goes wrong during the execution of any methods
      # which use the tar command internally.
      class Error < StandardError; end

      # Raised if something goes wrong during the Tar#compress_archive or
      # Tar#uncompress_archive methods.      
      class CompressError < StandardError; end

      # This class encapsulates tar & zip operations.
      class Tar::External
         # The version of the archive-tar-external library.
         VERSION = '1.2.3'

         # The name of the archive file to be used, e.g. "test.tar"
         attr_accessor :archive_name

         # The name of the tar program you wish to use.  The default is "tar".
         attr_accessor :tar_program

         # The name of the archive file after compression, e.g. "test.tar.gz"
         attr_reader :compressed_archive_name

         # Returns an Archive::Tar::External object.  The +archive_name+ is the
         # name of the tarball.  While a .tar extension is recommended based on
         # years of convention, it is not enforced.
         #
         # Note that this does not actually create the archive unless you
         # pass a value to +file_pattern+.  This then becomes a shortcut for
         # Archive::Tar::External.new + Archive::Tar::External#create_archive.
         #
         # If +program+ is provided, then it compresses the archive as well by
         # calling Archive::Tar::External#compress_archive internally.
         #
         def initialize(archive_name, file_pattern=nil, program=nil)
            @archive_name            = archive_name.to_s
            @compressed_archive_name = nil
            @tar_program             = 'tar'

            if file_pattern
               create_archive(file_pattern)
            end

            if program
               compress_archive(program)
            end
         end

         # Assign a compressed archive name.  This autogenerates the archive_name
         # based on the extension of the name provided, unless you provide the
         # extension yourself.  If the extension is '.tgz', then the base of the
         # name + '.tar' will be the new archive name.
         #
         # This should only be used if you have a pre-existing, compressed archive
         # that you want to uncompress, and want to have a Tar::External object
         # around.  Otherwise, use the class method Tar::External.uncompress.
         #
         def compressed_archive_name=(name, ext=File.extname(name))
            if ext.downcase == '.tgz'
               @archive_name = File.basename(name, ext.downcase) + '.tar'
            else
               @archive_name = File.basename(name, ext)
            end
            @compressed_archive_name = name
         end

         # Creates the archive using +file_pattern+.  Any errors that occur
         # here will raise a Error.
         #
         def create_archive(file_pattern)
            cmd = "#{@tar_program} cf #{@archive_name} #{file_pattern}"

            Open3.popen3(cmd){ |tar_in, tar_out, tar_err|
               err = tar_err.gets
               if err
                  raise Error, err.chomp
               end
            }

            self
         end

         alias :create :create_archive

         # Compresses the archive with +program+, or gzip if no program is
         # provided.  If you want to pass arguments to +program+, merely include
         # them as part of the program name, e.g. "gzip -f".
         #
         # Any errors that occur here will raise a Tar::CompressError.
         #
         def compress_archive(program='gzip')
            cmd = "#{program} #{@archive_name}"

            Open3.popen3(cmd){ |prog_in, prog_out, prog_err|
               err = prog_err.gets 
               if err
                  raise CompressError, err.chomp
               end

               # Find the new file name with the extension.  There's probably a more
               # reliable way to do this, but this should work 99% of the time.
               name = Dir["#{@archive_name}.{gz,bz2,cpio,zip}"].first
               @compressed_archive_name = name
            }

            self
         end

         alias :compress :compress_archive

         # Uncompresses the tarball using the program you pass to this method.  The
         # default is "gunzip".  Just as for +compress_archive+, you can pass
         # arguments along as part of the argument.
         #
         # Note that this is only for use with archives that have been zipped up
         # with gunzip, or whatever.  If you want to *extract* the files from the
         # tarball, use Tar::External#extract instead.
         #
         # Any errors that occur here will raise a Tar::CompressError.
         #
         def uncompress_archive(program="gunzip")
            unless @compressed_archive_name
               raise CompressError, "no compressed file found"
            end
               
            cmd = "#{program} #{@compressed_archive_name}"

            Open3.popen3(cmd){ |prog_in, prog_out, prog_err|
               err = prog_err.gets
               if err
                  raise CompressError, err.chomp
               end
               @compressed_archive_name = nil
            }
            self
         end

         alias :uncompress :uncompress_archive

         # Uncompress an existing archive, using +program+ to uncompress it.
         # The default decompression program is gunzip.
         #
         def self.uncompress_archive(archive, program='gunzip')
            cmd = "#{program} #{archive}"

            Open3.popen3(cmd){ |prog_in, prog_out, prog_err|
               err = prog_err.gets
               if err
                  raise CompressError, err.chomp
               end
            }
         end

         class << self
            alias uncompress uncompress_archive
         end

         # An alias for Tar::External.uncompress_archive.
         #
         #def self.uncompress(file, program='gunzip')
         #   self.uncompress_archive(file, program)
         #end

         # Returns an array of file names that are included within the tarball.
         # This method does not extract the archive.
         #
         def archive_info
            result = []
            cmd = "#{@tar_program} tf #{@archive_name}"
            Open3.popen3(cmd){ |ain, aout, aerr|
               err = aerr.gets
               if err
                  raise Error, err.chomp
               end

               while output = aout.gets
                  result << output.chomp
               end
            }
            result
         end

         alias :info :archive_info

         # Adds +files+ to an already existing archive.
         #
         def add_to_archive(*files)
            if files.empty?
               raise Error, "there must be at least one file specified"
            end

            cmd = "#{@tar_program} rf #{@archive_name} #{files.join(" ")}"
            Open3.popen3(cmd){ |ain, aout, aerr|
               err = aerr.gets

               if err
                  raise Error, err.chomp
               end
            }
            self
         end

         alias :add :add_to_archive

         # Updates the given +files+ in the archive, i.e. they are added if they
         # are not already in the archive or have been modified.
         #
         def update_archive(*files)
            if files.empty?
               raise Error, "there must be at least one file specified"
            end

            cmd = "#{@tar_program} uf #{@archive_name} #{files.join(" ")}"

            Open3.popen3(cmd){ |ain, aout, aerr|
               err = aerr.gets
               if err
                  raise Error, err.chomp
               end
            }

            self
         end

         alias :update :update_archive

         # Expands the contents of the tarball.  It does NOT delete the tarball.
         # If +files+ are provided, then only those files are extracted.
         # Otherwise, all files are extracted.
         #
         # Note that some tar programs, notably the tar program shipped by Sun,
         # does not issue any sort of warning or error if you try to extract a
         # file that does not exist in the archive.
         #
         def extract_archive(*files)
            cmd = "#{@tar_program} xf #{@archive_name}"

            unless files.empty?
               cmd << " " << files.join(" ")
            end

            Open3.popen3(cmd){ |ain, aout, aerr|
               err = aerr.gets

               if err
                  raise Error, err.chomp
               end
            }

            self
         end

         alias :expand_archive :extract_archive
         alias :extract :extract_archive
         alias :expand :extract_archive

         # A class method that behaves identically to the equivalent instance
         # method, except that you must specifiy that tarball as the first
         # argument. Also, the tar program is hard coded to 'tar xf'.
         #
         def self.extract_archive(archive, *files)
            cmd = "tar xf #{archive}"

            unless files.empty?
               cmd << " " << files.join(" ")
            end

            Open3.popen3(cmd){ |ain, aout, aerr|
               err = aerr.gets

               if err
                  raise Error, err.chomp
               end
            }

            self
         end

         class << self
            alias expand_archive extract_archive
            alias extract extract_archive
            alias expand extract_archive
         end
      end
   end
end
